Automattic/sensei
# Learner mode settings

Fixes #4510
Fixes #4511
Fixes #4555

### Changes proposed in this Pull Request

* Added basic support for "deep linking" for Sensei Settings page. With this change, if you append section id into the Sensei Settings page url then it will select that section tab. For example: `/wp-admin/admin.php?page=sensei-settings#course-settings` will pre-select (pre-open) the courses section (tab) of the settings.
* Added `Sensei_Course_Theme::get_sensei_theme_customize_url()` method that generates a url that allows admins/teachers to customize the course theme styles. This basically gets the customize url for the latest lesson that is assigned to any course. If that course has no learner mode enabled, then the url opens in preview mode. This one probably needs some decision from the product. Which page should be opened for customizing Learner mode styles? Is it the latest edited Lesson post? Is it the latest edited Course post?
* Added the global **Learner mode** setting. Which overwrites the **Learner mode** setting for the individual courses. We still need design for learner mode setting for the individual course editor page when the global learner mode is enabled. For now it simply says "Learner mode is enabled globally". Which links to the global setting.
* Sensei settings are now exposed into js runtime on the admin. This allows us to detect if the **Learner Mode** is enabled or not on the Course editor page. Maybe also useful in other parts of Sensei.

### Testing instructions


* Enable the feature flag: `add_filter( 'sensei_feature_flag_course_theme', '__return_true' );`
* Edit any Course and confirm that you can turn on/off the **Learner Mode**. (see video below)
* Confirm that you can preview the course in Learner Mode when Learner mode is enabled for that course.
* Confirm that you can preview the course in Learner Mode when Learner mode is **not** enabled for that course.
* Confirm that you can customize the Learner Mode styles when Learner mode is enabled for that course.
* Confirm that you can customize the Learner Mode styles when Learner mode is **not** enabled for that course.
* In the Sensei Settings page, in Courses tab, confirm that you can enable/disable the Learner Mode for all courses.
* In the Sensei Settings page, in Courses tab, confirm that the "Customize Styles" allows you to customize Learner Mode styles.


### New/Updated Hooks

* `sensei_course_learning_mode_enabled` Filters if a course has learning mode enabled.
  * Parameters:
    * `{boolean} $enabled_for_course`. True if the learning mode is enabled for the course.
    * `{boolean} $enabled_globally`. True if the learning mode is enabled globally.
    * `{int} $course_id`. The id of the course.
  * Returns: `boolean`.



### Screenshot / Video


https://user-images.githubusercontent.com/2578542/147478201-72f66667-1674-4de3-87fc-59547e848212.mp4
